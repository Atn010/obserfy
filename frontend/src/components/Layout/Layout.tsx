import React, { FC, useState } from "react"
import { useColorMode } from "theme-ui"
import { navigate } from "gatsby"
import AppBar, { StoreName } from "../AppBar/AppBar"
import SideBar from "../SideBar/SideBar"
import { ReactComponent as HomeIcon } from "../../icons/home.svg"
import { ReactComponent as SettingsIcon } from "../../icons/settings.svg"
import { ReactComponent as LightModeIcon } from "../../icons/light-mode.svg"
import { ReactComponent as DarkModeIcon } from "../../icons/dark-mode.svg"
import NavigationItem from "../NavigationItem/NavigationItem"
import { Typography } from "../Typography/Typography"
import Spacer from "../Spacer/Spacer"
import Icon from "../Icon/Icon"
import Box from "../Box/Box"
import Flex, { FlexProps } from "../Flex/Flex"
import "../layout.css"
import { getSchoolId } from "../../hooks/schoolIdState"

/** Top level component that encapsulate every page and provides navigation and
 * everything */
interface Props {
  pageTitle: string
}

export const Layout: FC<Props> = ({ pageTitle, children }) => {
  const schoolId = getSchoolId()
  if (schoolId === null) {
    navigate("/choose-school")
  }

  const [isSidebarShown, setShowSidebar] = useState(false)
  const toggleSidebar = (): void => setShowSidebar(!isSidebarShown)

  return (
    <>
      <AppBar title={pageTitle} onMenuClick={toggleSidebar} position="fixed" />
      <MainSidebar closeSidebar={toggleSidebar} isShown={isSidebarShown} />
      <Box
        as="main"
        width="100%"
        mt="57px"
        pl={[0, 230]}
        backgroundColor="background"
      >
        {children}
      </Box>
    </>
  )
}

const MainSidebar: FC<{
  closeSidebar: () => void
  isShown: boolean
}> = ({ closeSidebar, isShown }) => {
  const [colorMode, setColorMode] = useColorMode()
  return (
    <SideBar isShown={isShown} onOutsideClick={closeSidebar}>
      <StoreName
        backgroundColor="surface"
        mb={3}
        alignItems="center"
        width="100%"
        opacity={[1, 0]}
        sx={{ boxShadow: "low" }}
      />
      <NavigationItem
        onClick={closeSidebar}
        to="/"
        text="Home"
        icon={HomeIcon}
      />
      <Spacer />
      <NavigationItem
        onClick={closeSidebar}
        to="/settings"
        text="settings"
        icon={SettingsIcon}
        mb={3}
      />
      <SidebarToggleItem
        onClick={() => {
          setColorMode(colorMode === "dark" ? "default" : "dark")
        }}
        text={colorMode === "dark" ? "Light Mode" : "Dark Mode"}
        icon={colorMode === "dark" ? LightModeIcon : DarkModeIcon}
        data-cy="toggleTheme"
      />
    </SideBar>
  )
}

interface SidebarToggleItemProps extends FlexProps {
  text: string
  icon?: FC
}

export const SidebarToggleItem: FC<SidebarToggleItemProps> = ({
  text,
  icon,
  ...props
}) => {
  return (
    // eslint-disable-next-line react/jsx-no-undef
    <Flex
      as="button"
      alignItems="center"
      m={2}
      my={0}
      pl={2}
      sx={{
        borderStyle: "none",
        cursor: "pointer",
        borderRadius: "default",
        textDecoration: "none",
        backgroundColor: "transparent",
        "&:hover": {
          backgroundColor: "primaryLightest",
        },
      }}
      {...props}
    >
      <Icon
        alt={`${text} icon`}
        ml={0}
        my={2}
        as={icon}
        sx={{ fill: "textMediumEmphasis" }}
      />
      <Typography.Body m={0} fontSize={1} color="textMediumEmphasis">
        {text}
      </Typography.Body>
    </Flex>
  )
}

export default Layout
