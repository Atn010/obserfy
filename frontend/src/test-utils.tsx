/* eslint-disable import/no-extraneous-dependencies */
import React, { FC, ReactElement } from "react"
import { ThemeProvider } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import { IntlProvider } from "gatsby-plugin-intl3"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import Theme from "./gatsby-plugin-theme-ui"
import Box from "./components/Box/Box"
import en from "./intl/en.json"

const AllProviders: FC = ({ children }) => {
  // TODO: Maybe this better be outside
  const history = createHistory(createMemorySource("/"))
  return (
    <LocationProvider history={history}>
      <ThemeProvider theme={Theme}>
        <IntlProvider locale="en" messages={en}>
          <Box fontSize={[16, 20]}>{children}</Box>
        </IntlProvider>
      </ThemeProvider>
    </LocationProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
