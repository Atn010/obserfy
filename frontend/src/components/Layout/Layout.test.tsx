import React from "react"
import { render } from "../../test-utils"
import { Basic } from "./Layout.stories"

describe("Layout", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
