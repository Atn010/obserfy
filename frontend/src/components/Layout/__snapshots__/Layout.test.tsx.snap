// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render correctly 1`] = `
.emotion-14 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-14 {
    font-size: 20px;
  }
}

.emotion-12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding: 8px;
  background-color: var(--theme-ui-colors-background,#F7F8FA);
  border-radius: 8px;
  box-shadow: rgba(0,0,0,0.1) 0px 0px 1px,rgba(0,0,0,0.27) 0px 2px 4px -2px;
  z-index: 5;
  top: auto;
  bottom: 0;
  left: 0;
  position: fixed;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  border-radius: 0;
  padding: 0;
  background-color: var(--theme-ui-colors-surfaceTransparent,rgba(255,255,255,0.965));
  overflow: hidden;
  display: block;
  height: auto;
  width: 100%;
  padding-top: 0;
}

@media screen and (min-width:40em) {
  .emotion-12 {
    top: 0;
    bottom: auto;
  }
}

@media screen and (min-width:40em) {
  .emotion-12 {
    display: block;
    height: 100%;
    width: 70px;
    padding-top: 8px;
  }
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  height: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media screen and (min-width:40em) {
  .emotion-11 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    height: 100%;
  }
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  height: 70px;
  width: 70px;
  display: none;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    display: block;
  }
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  height: 48px;
  margin-bottom: 0;
  width: 60px;
  padding-top: 4px;
  padding-bottom: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-3:after {
  top: 0;
  right: inherit;
  position: absolute;
  background-color: var(--theme-ui-colors-textPrimary,#00664a);
  border-radius: 0 0 10px 10px;
  width: 0%;
  height: 2px;
  content: '';
  -webkit-transition: width 100ms cubic-bezier(0.0,0.0,0.2,1);
  transition: width 100ms cubic-bezier(0.0,0.0,0.2,1);
}

@media screen and (min-width:40em) {
  .emotion-3:after {
    top: inherit;
    right: 0;
    border-radius: 10px 0 0 10px;
    width: 2px;
    height: 0%;
  }
}

@media screen and (min-width:40em) {
  .emotion-3 {
    height: 60px;
    margin-bottom: 8px;
    width: 70px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-textDisabled,rgba(0,0,0,0.38));
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0;
  height: 24px;
  width: 24px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 10px;
  line-height: 1;
  font-weight: normal;
}

@media screen and (min-width:40em) {
  .emotion-2 {
    font-size: 12px;
  }
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  height: 100%;
  display: none;
}

@media screen and (min-width:40em) {
  .emotion-7 {
    display: block;
  }
}

.emotion-13 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  padding-left: 0;
  padding-bottom: 48px;
  background-color: var(--theme-ui-colors-background,#F7F8FA);
}

@media screen and (min-width:40em) {
  .emotion-13 {
    padding-left: 80px;
    padding-bottom: 0;
  }
}

<div>
  <div
    class="emotion-14"
  >
    <nav
      class="emotion-12"
    >
      <div
        class="emotion-11"
      >
        <div
          class="emotion-0"
        />
        <a
          href="/dashboard/home"
          style="outline: none;"
        >
          <div
            class="emotion-3"
          >
            <div
              class="emotion-1"
            />
            <p
              class="emotion-2"
            >
              Observe
            </p>
          </div>
        </a>
        <a
          href="/dashboard/analyze"
          style="outline: none;"
        >
          <div
            class="emotion-3"
          >
            <div
              class="emotion-1"
            />
            <p
              class="emotion-2"
            >
              Analyze
            </p>
          </div>
        </a>
        <div
          class="emotion-7"
        />
        <a
          href="/dashboard/settings"
          style="outline: none;"
        >
          <div
            class="emotion-3"
          >
            <div
              class="emotion-1"
            />
            <p
              class="emotion-2"
            >
              Settings
            </p>
          </div>
        </a>
      </div>
    </nav>
    <main
      class="emotion-13"
    />
  </div>
</div>
`;
